cmd/kubeadm/app/cmd/phases/reset/unmount_linux.go

```diff
@@ -32,6 +32,11 @@ func unmountKubeletDirectory(absoluteKubeletRunDirectory string) error {
 	if err != nil {
 		return err
 	}
+
+	if !strings.HasSuffix(absoluteKubeletRunDirectory, "/") {
+		absoluteKubeletRunDirectory += "/"
+	}
+
```

Comment:
i'm trying to understand what happens here.

this call to EvalSymlinks results in a path that does not end in /
https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/cmd/kubeadm/app/cmd/phases/reset/cleanupnode.go#L97

when iterating the lines in /proc/mounts we find an entry that is `device /var/lib/kubelet`, thus it ends up being unmounted as it is not skipped at this line:

```
if len(m) < 2 || !strings.HasPrefix(m[1], absoluteKubeletRunDirectory) {
```

your change make the line if condition above skip the block device mount of /var/lib/kubelet?

Reply:
Thanks for review, please correct me if I am wrong here, this was an attempt to replicate the following fixes https://github.com/kubernetes/kubernetes/pull/71663/files

Here is the sample test case from the above code snippets:
```
# mount | grep kubelet
/dev/sdc1 on /var/lib/kubelet type ext4 (rw,relatime,seclabel,data=ordered)
# awk '$2 ~ path {print $2}' path=/var/lib/kubelet/ /proc/mounts <== PR #71663 fixes this by adding a "/"
# awk '$2 ~ path {print $2}' path=/var/lib/kubelet /proc/mounts
/var/lib/kubelet
```
So in short, yes it will skip the `/var/lib/kubelet` directory during the loop

Reply:
thanks for the confirmation.

Reply:
> i'm trying to understand what happens here.
> 
> this call to EvalSymlinks results in a path that does not end in /

Let's say that it's not guaranteed (especially if `kubeadmconstants.KubeletRunDirectory` is not a symlink). We shouldn't rely on `EvalSymlinks` to return a result with a trailing slash.

> https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/cmd/kubeadm/app/cmd/phases/reset/cleanupnode.go#L97
> 
> when iterating the lines in /proc/mounts we find an entry that is `device /var/lib/kubelet`, thus it ends up being unmounted as it is not skipped here:
> 
> ```
> if len(m) < 2 || !strings.HasPrefix(m[1], absoluteKubeletRunDirectory) {
> ```
> 
> your change makes the if condition above skip the block device mount of /var/lib/kubelet?

Yes. The paths in `/proc/mounts` don't have trailing slashes so adding an extra slash to `absoluteKubeletRunDirectory` would save the day.

Reply:
@rosti @thtanaka @neolit123 

> We shouldn't rely on EvalSymlinks to return a result with a trailing slash.

According to the [official golang documentation](https://golang.org/pkg/path/filepath/#EvalSymlinks) EvalSymlinks calls filepath.Clean on the result.
[Clean docs state that](https://golang.org/pkg/path/filepath/#Clean) ```The returned path ends in a slash only if it represents a root directory```

So, to my understanding, we can rely on that.